lib/monitoring/middleware/prometheus_collector.rb

```diff
@@ -0,0 +1,76 @@
+require 'benchmark'
+require_relative '../operations.rb'
+
+module Monitoring
+  module Middleware
+    class PrometheusCollector
+      attr_reader :app
+
+      def initialize(app, options = {})
+        @app = app
+        @pubsub = options[:pubsub]
+      end
+
+      def call(env) # :nodoc:
+        trace(env) { @app.call(env) }
+      end
+
+      protected
+
+      # Trace HTTP requests
+      def trace(env)
+        response = nil
+        duration = Benchmark.realtime { response = yield }
+        record(env, response.first.to_s, duration)
+        return response
+      rescue => exception
+        method = env['REQUEST_METHOD']
+        path = strip_ids_from_path(env['PATH_INFO'])
+        operation = find_operation(method, path)
+        @pubsub.publish(
+          "conjur.request_exception", 
+          operation: operation,
+          exception: exception
+        )
+        raise
+      end
+
+      def record(env, code, duration)
+        method = env['REQUEST_METHOD']
+        path = strip_ids_from_path(env['PATH_INFO'])
+        operation = find_operation(method, path)
+
+        @pubsub.publish(
+          "conjur.request", 
+          code: code,
+          operation: operation
+        )
+
+        @pubsub.publish(
+          "conjur.request_duration", 
+          duration: duration,
+          operation: operation
+        )
+      rescue
+        # TODO: log unexpected exception during request recording
+        nil
+      end
+
+      def strip_ids_from_path(path)
```

Comment:
This is something that the solution design unfortunately doesn't cover.

Do we still need this? For `find_operation`, if we get a match we don't care about the path and if we don't get a match then I want to think the path isn't our responsibility. That last part feels a bit negligent so maybe its best to just have the operation be `unknown`, since even if we decided to take on the responsibility the possibilities of paths are endless.

Reply:
Good question. To me, it seems helpful in the current implementation where we are fetching the operation from a file where there is no guarantee that it will always be in sync with the API spec. But if we are able to utilize something dynamic like `Rails.application.routes.recognize_path` then I think that need goes away. I'll do some investigating